### PR TITLE
Move calculation of metrics to `dm_pix`

### PR DESCRIPTION
**Pull request by phinate**
Overhauls our metric calculations to use JAX and [DeepMind's `pix` library](https://github.com/google-deepmind/dm_pix). Working from a pipeline perspective, which is great!

Issues right now:
- SSIM is producing NaNs (I've added a test to verify this)
- I've [forked the repo](https://github.com/phinate/dm_pix/tree/ignore-nans) to replace mean -> nanmean, but I haven't worked out how to fix this for SSIM. Will require an explicit install:
```
pip install git+https://github.com/phinate/dm_pix@ignore-nans
```
or ideally, just clone my fork and make some edits if you want to help debug this. I've added @dfulu as a collaborator since he's been working on metrics the most with me, but any help is appreciated!

### CODE REVIEW

**Comment by phinate** [comment]
Problems are solved, so have marked this ready for review.

Still need to add install instructions for the fork of `dm_pix`. We could possibly add it as a submodule. Open to thoughts?

**Comment by dfulu** [comment]
> Problems are solved, so have marked this ready for review.
> 
> Still need to add install instructions for the fork of `dm_pix`. We could possibly add it as a submodule. Open to thoughts?

From what I can tell, it seems like we are only dependent on this [one module of dm_pix](https://github.com/phinate/dm_pix/blob/ignore-nans/dm_pix/_src/metrics.py) and its only around 250 lines of code. Should we just copy that into this repo rather than remaining dependent on that external fork?

**Comment by phinate** [comment]
> Should we just copy that into this repo rather than remaining dependent on that external fork?

I also thought this. I guess I didn't know if we'd want to use more functionality from the lib. Happy to take that approach for now, since it works nicely for the medium-term.

**Comment by phinate** [comment]
Have added the relevant code to our repo, keeping the name `metrics.py`, but importing as `dm_pix` where used for forward compatibility.

TL;DR on what the problems were: our tests failed because the filter size for the convolution in `metrics.ssim` was bigger than the image; the nan in the top-left corner made the whole image be replaced with nans. This is solved by just reducing the filter size during tests (our images there are only 8x9 to speed up computations, so our filter size needs to be around 9 or less, instead of the default of 11).

File: README.md [hunk 1]
```diff
@@ -8,7 +8,20 @@ Tooling and infrastructure to enable cloud nowcasting.
...
+```zsh
+git clone https://github.com/alan-turing-institute/cloudcasting
+cd cloudcasting
+python -m pip install .
+```
+
+To run metrics on GPU:
+
+```zsh
+python -m install --upgrade "jax[cuda12]"
```
Comment (dfulu) [inline README.md:22]: Does this mean by default the calculations will be run on the GPU if available?

Comment (phinate) [reply README.md:22]: Yes, `jax.default_backend()` should default to `'gpu'`. The strict check would be to change messaging when running validation if this returned `'cpu'`.

Comment (phinate) [reply README.md:22]: A sharp edge here is if we happen to have tensorflow installed in our HPC env, where it will take priority access of GPUs. The workaround is 
```python
import tensorflow as tf
tf.config.experimental.set_visible_devices([], "GPU")
```
but we may not run into this!

**Review by dfulu — APPROVED** [review]

File: src/cloudcasting/validation.py [hunk 2]
```diff
@@ -183,13 +196,38 @@ def score_model_on_all_metrics(
...
-        "mae": mae_batch,
-        "mse": mse_batch,
-        # "ssim": ssim_batch,  # currently unstable with nans
-    }
+    if metric_kwargs is None:
+        metric_kwargs_dict: dict[str, dict[str, Any]] = {}
+    else:
+        metric_kwargs_dict = metric_kwargs
+
+    def get_pix_function(
```
[the remark this answers is not in the document]
Comment (phinate) [reply src/cloudcasting/validation.py:204]: For documentation purposes: this is a way to fish functions from `dm_pix` by name, which is a little bit over-engineered since we now have the metric code in our library. Will re-visit at a later stage once things are more mature.
